File: src/bridge/text.py
from __future__ import annotations

import hashlib
import re

from .models import BridgeEvent, EventKind, Platform

_MM_MARKDOWN = re.compile(r"([\\`*_{}\[\]()<>#+\-.!|~])")
_MM_MENTION = re.compile(r"@(?=[\w.-])", re.UNICODE)
_AUTHOR_MARKERS = ("🔵", "🟢", "🟣", "🟠", "🟡", "🟤", "⚫")


def author_marker(event: BridgeEvent) -> str:
    identity = f"{event.platform.value}:{event.author_id}".encode()
    return _AUTHOR_MARKERS[hashlib.sha256(identity).digest()[0] % len(_AUTHOR_MARKERS)]


def author_label(event: BridgeEvent) -> str:
    suffix = "Mattermost" if event.platform is Platform.MATTERMOST else "Telegram"
    username = f" (@{event.author_username})" if event.author_username else ""
    return f"{author_marker(event)} {event.author_name} · {suffix}{username}"


def render_text(
    event: BridgeEvent,
    target: Platform,
    unknown_quote: str | None = None,
    *,
    include_author: bool = True,
) -> str:
    if event.kind is EventKind.EDIT:
        marker = "✏️ Исправление"
    elif event.kind is EventKind.DELETE:
        marker = "🗑️ Сообщение удалено"
    else:
        marker = ""

    body = event.text.strip()
    if unknown_quote:
        quote = unknown_quote.replace("\n", " ").strip()[:240]
        body = f"↩ Ответ на несвязанное сообщение: «{quote}»\n\n{body}".strip()
    if marker:
        body = f"{marker}: {body}".rstrip(": ")
    if include_author and event.platform is Platform.MATTERMOST and target is Platform.TELEGRAM:
        result = f"{event.author_name} : {body}".rstrip(" :")
    elif include_author and event.platform is Platform.TELEGRAM and target is Platform.MATTERMOST:
        result = f"{event.author_name}: {body}".rstrip(": ")
    elif include_author:
        result = f"{author_label(event)}\n{body}".rstrip()
    else:
        result = body

    if target is Platform.MATTERMOST:
        # External text must not create Mattermost formatting or notifications.
        result = _MM_MARKDOWN.sub(r"\\\1", result)
        result = _MM_MENTION.sub("@\u200b", result)
    return result


def split_text(text: str, limit: int) -> list[str]:
    if len(text) <= limit:
        return [text]
    # Reserve room for a stable (N/M) prefix. Recompute if total grows to 3 digits.
    chunk_size = max(1, limit - 12)
    chunks: list[str] = []
    remaining = text
    while remaining:
        if len(remaining) <= chunk_size:
            chunks.append(remaining)
            break
        cut = remaining.rfind("\n", 0, chunk_size + 1)
        if cut < chunk_size // 2:
            cut = remaining.rfind(" ", 0, chunk_size + 1)
        if cut < chunk_size // 2:
            cut = chunk_size
        chunks.append(remaining[:cut].rstrip())
        remaining = remaining[cut:].lstrip()
    total = len(chunks)
    return [f"({index}/{total}) {chunk}" for index, chunk in enumerate(chunks, 1)]


def safe_filename(name: str) -> str:
    cleaned = name.replace("/", "_").replace("\\", "_").replace("\x00", "")
    cleaned = re.sub(r"[\r\n\t]+", " ", cleaned).strip()
    return cleaned[:240] or "attachment"

File: internal/textutil/text.go
package textutil

import (
	"crypto/sha256"
	"fmt"
	"regexp"
	"strings"
	"unicode/utf8"

	"github.com/elliotvegaagent/telegram-mattermost-bridge/internal/model"
)

var (
	mmMarkdown  = regexp.MustCompile("([\\\\`*_{}\\[\\]()<>#+\\-.!|~])")
	mmMention   = regexp.MustCompile(`@([\pL\pN_.-])`)
	badFilename = regexp.MustCompile(`[\r\n\t]+`)
)

var markers = []string{"🔵", "🟢", "🟣", "🟠", "🟡", "🟤", "⚫"}

func AuthorMarker(event model.Event) string {
	sum := sha256.Sum256([]byte(string(event.Platform) + ":" + event.AuthorID))
	return markers[int(sum[0])%len(markers)]
}

func AuthorLabel(event model.Event) string {
	suffix := "Telegram"
	if event.Platform == model.Mattermost {
		suffix = "Mattermost"
	}
	username := ""
	if event.AuthorUsername != "" {
		username = " (@" + event.AuthorUsername + ")"
	}
	return fmt.Sprintf("%s %s · %s%s", AuthorMarker(event), event.AuthorName, suffix, username)
}

func Render(event model.Event, target model.Platform, unknownQuote string, includeAuthor bool) string {
	marker := ""
	if event.Kind == model.Edit {
		marker = "✏️ Исправление"
	}
	if event.Kind == model.Delete {
		marker = "🗑️ Сообщение удалено"
	}
	body := strings.TrimSpace(event.Text)
	if unknownQuote != "" {
		quote := strings.TrimSpace(strings.ReplaceAll(unknownQuote, "\n", " "))
		quote = truncateRunes(quote, 240)
		body = strings.TrimSpace(fmt.Sprintf("↩ Ответ на несвязанное сообщение: «%s»\n\n%s", quote, body))
	}
	if marker != "" {
		body = strings.TrimRight(marker+": "+body, ": ")
	}
	var result string
	if includeAuthor && event.Platform == model.Mattermost && target == model.Telegram {
		result = strings.TrimSpace(event.AuthorName + "\n" + body)
	} else if includeAuthor && event.Platform == model.Telegram && target == model.Mattermost {
		result = strings.TrimRight(event.AuthorName+": "+body, ": ")
	} else if includeAuthor {
		result = strings.TrimSpace(AuthorLabel(event) + "\n" + body)
	} else {
		result = body
	}
	if target == model.Mattermost {
		result = mmMarkdown.ReplaceAllString(result, `\$1`)
		result = mmMention.ReplaceAllString(result, "@\u200b$1")
	}
	return result
}

func Split(value string, limit int) []string {
	if utf8.RuneCountInString(value) <= limit {
		return []string{value}
	}
	runes := []rune(value)
	chunkSize := limit - 12
	if chunkSize < 1 {
		chunkSize = 1
	}
	chunks := []string{}
	for len(runes) > 0 {
		if len(runes) <= chunkSize {
			chunks = append(chunks, string(runes))
			break
		}
		cut := chunkSize
		for i := chunkSize; i >= chunkSize/2; i-- {
			if runes[i-1] == '\n' {
				cut = i - 1
				break
			}
		}
		if cut == chunkSize {
			for i := chunkSize; i >= chunkSize/2; i-- {
				if runes[i-1] == ' ' {
					cut = i - 1
					break
				}
			}
		}
		if cut < 1 {
			cut = chunkSize
		}
		chunks = append(chunks, strings.TrimRight(string(runes[:cut]), " \n\t"))
		runes = []rune(strings.TrimLeft(string(runes[cut:]), " \n\t"))
	}
	result := make([]string, len(chunks))
	for i, chunk := range chunks {
		result[i] = fmt.Sprintf("(%d/%d) %s", i+1, len(chunks), chunk)
	}
	return result
}

func SafeFilename(name string) string {
	name = strings.ReplaceAll(strings.ReplaceAll(strings.ReplaceAll(name, "/", "_"), "\\", "_"), "\x00", "")
	name = strings.TrimSpace(badFilename.ReplaceAllString(name, " "))
	name = truncateRunes(name, 240)
	if name == "" {
		return "attachment"
	}
	return name
}

func truncateRunes(value string, limit int) string {
	runes := []rune(value)
	if len(runes) <= limit {
		return value
	}
	return string(runes[:limit])
}
